build.wake

```diff
@@ -0,0 +1,416 @@
+# some region of memory.
+export def freedomESDKProgramOverflowsMemory sdkTarget program =
+  def programJob = makeProgramJob sdkTarget program
+  def jobStderr = programJob.runJob.getJobStderr | getWhenFail ""
+  def sectionWillNotFit = `.*ld: .* section .* will not fit in region.*`
+  def regionOverflowed = `.*ld: region .* overflowed by .* bytes.*`
+
+  (matches sectionWillNotFit jobStderr) && (matches regionOverflowed jobStderr)
+
+# This rule can be used to detect when the build of a FreedomESDKProgram
+# fails for a given FreedomESDKTarget as a result of the memory layout
+# of the Target interacting with the default Freedom E SDK linker
+# scripts in a way which is incompatible with the available code models
+# in the RISC-V toolchain.
+export def freedomESDKProgramCannotLinkUnderCodeModel sdkTarget program =
+  def programJob = makeProgramJob sdkTarget program
+  def jobStderr = programJob.runJob.getJobStderr | getWhenFail ""
+  def relocationTruncated = `.*relocation truncated to fit.*`
+
+  matches relocationTruncated jobStderr
+
+target makeProgram sdkTarget program =
+  def targetName = sdkTarget.getFreedomESDKTargetName
+
+  def outputs =
+    makeProgramJob sdkTarget program
+    | runJob
+    | getJobOutputs
+
+  def filterFiles regex = filter (matches regex _.getPathName) outputs
+  def getFile regex = filterFiles regex | head
+  def elfResult = getFile `.*\.elf`
+  def hexResult = getFile `.*\.hex`
+
+  match elfResult hexResult
+    (Some elf) (Some hex) = Pair elf hex
+    _ _ =
+      def badPath =
+        "{targetName}: Unexpected software ouputs: {format outputs}"
+        .makeError
+        .makeBadPath
+      Pair badPath badPath
+
+def makeProgramJob sdkTarget program =
+  def sdk = sdkTarget.getFreedomESDKTargetSDK
+  def sdkDir = sdk.getFreedomESDKDir
+  def buildEnv = sdk.getFreedomESDKBuildEnv
+  def targetName = sdkTarget.getFreedomESDKTargetName
+  def programName = program.getFreedomESDKProgramName
+
+  def cmdline =
+    def venvPath = freedomESDKVenvDir buildEnv
+    "bash", "-c", "%
+      export FREEDOM_E_SDK_VENV_PATH=$(pwd)/%{venvPath}
+      export FREERTOS_METAL_VENV_PATH=$(pwd)/%{venvPath}
+      touch ${FREEDOM_E_SDK_VENV_PATH}/.stamp
+      make -C %{sdkDir} software \
+        ARFLAGS=Dcr \
+        CONFIGURATION=release \
+        TARGET=%{targetName} \
+        PROGRAM=%{programName}
+      %", Nil
+
+  def inputs =
+    def sdkInstall = sdk.getFreedomESDKInstalledFiles
+    def targetInstall = sdkTarget.getFreedomESDKTargetInstalledFiles
+    def programInstall = installFreedomESDKProgram sdk program
+    sdkInstall
+    ++ targetInstall
+    ++ programInstall
+    ++ (installFreedomESDKVirtualenv buildEnv).getFreedomESDKVirtualenvInstalledFiles
+
+  def fnOutputs _ = files "{sdkDir}/software/{programName}/release" `[^/]*.(elf|hex)`
+
+  makePlan cmdline inputs
+  | setPlanResources buildEnv.getFreedomESDKBuildEnvResources
+  | setPlanFnOutputs fnOutputs
+  | setPlanLocalOnly True
```

Comment:
Once again, do you know why this requires LocalOnly?

Reply:
Not sure, I think it's carried over from the original wake for this task. I suppose I'll get rid of it and find out if it's necessary.